### DISCUSSION TITLE
JDK 17 Tests

### DISCUSSION THREAD

**Discussion by lorenzo-gomez-windhover** [ControlSystemStudio/phoebus#2077] [General]
Thought I would start a conversation to share progress/ideas when testing jdk17.

**Comment by lorenzo-gomez-windhover** [comment]
Phoebus seems to compile and run fine under the following conditions:

- [Maven>=3.8.4](https://dlcdn.apache.org/maven/maven-3/)
- [JDK17](https://jdk.java.net/17/)

Here are the details of the OS where ran tests:
```
NAME="Ubuntu"
VERSION="18.04.6 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.6 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic

```

I suspect that this applies to other platforms as well(Windows, MacOS, etc), but we should probably test those to verify. I'll let you all know of any other progress/developments if there is any in the future as I will try to daily drive jdk17 on the daily while developing phoebus.

**Comment by georgweiss** [comment]
Another factor to consider is language features. I would be interested in the ``record`` keyword, available from Java14. Currently we set the Java "language level" in the top level pom, maven-compiler-plugin configuration. What happens if this is changed to 17?

**Reply by lorenzo-gomez-windhover** [reply]
I see the `source` and `target` language levels you are talking about...When I change those to 17, phoebus does not compile.
The following modules fail to compile when the language level is set to 17:

- `core-vtype 4.6.6-SNAPSHOT`
- `core-util 4.6.6-SNAPSHOT`
- `core-pva 4.6.6-SNAPSHOT`
- `app-save-and-restore-model 4.6.6-SNAPSHOT `

Here is output from maven:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.2:compile (default-compile) on project core-pva: Compilation failure -> [Help 1]
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.2:compile (default-compile) on project core-vtype: Compilation failure -> [Help 1]
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.2:compile (default-compile) on project core-util: Compilation failure -> [Help 1]
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.2:compile (default-compile) on project app-save-and-restore-model: Compilation failure -> [Help 1]

```
We definitely have to investigate this further.

**Comment by kasemir** [comment]
While you're at it, can you copy some more detail for the "Compilation failure"? There should be some more specific error including source file, line, reason.

**Comment by georgweiss** [comment]
To make it compile using source and target 17, you need to do the following:
1) Install jdk 17 and make sure that is used by Maven. Check with ``mvn -version`` to verify that Maven is using jdk 17.
2) In pom.xml in services/alarm-config-logger, services/alarm-logger and services/save-and-restore: add ``<configuration><mainClass>TheMainClass</mainClass></configuration>``  for the spring-boot-plugin. I also updated to version 2.4.1 of the plugin.
3) Skip tests when building: ``mvn -DskipTests=true clean install`` 

To build with unit tests more work is required, e.g. eliminate Lombok (which is why I want the ``record`` keyword).

Once built, Phoebus seems to run fine. Services not tested, though.
